### PR TITLE
Fix InvalidURIError with git ssh remotes

### LINKED ISSUE #342: As of yesterday Addressible broke `travis`
```
System
Ruby:                     Ruby 2.2.3-p173
Operating System:         Ubuntu 14.04
RubyGems:                 RubyGems 2.5.0

CLI
Version:                  1.8.0
Plugins:                  none
Auto-Completion:          yes
Last Version Check:       2015-12-08 13:08:19 -0600

Session
API Endpoint:             https://api.travis-ci.org/
Logged In:                as "envygeeks"
Verify SSL:               yes
Enterprise:               no

Endpoints
org:                      https://api.travis-ci.org/ (access token, current)

Last Exception
An error occurred running `travis encrypt`:
    Addressable::URI::InvalidURIError: Invalid scheme format: git@github.com
        from /usr/local/lib/ruby/gems/2.2.0/gems/addressable-2.4.0/lib/addressable/uri.rb:867:in `scheme='
        from /usr/local/lib/ruby/gems/2.2.0/gems/addressable-2.4.0/lib/addressable/uri.rb:795:in `block in initialize'
        from /usr/local/lib/ruby/gems/2.2.0/gems/addressable-2.4.0/lib/addressable/uri.rb:2302:in `call'
        from /usr/local/lib/ruby/gems/2.2.0/gems/addressable-2.4.0/lib/addressable/uri.rb:2302:in `defer_validation'
        from /usr/local/lib/ruby/gems/2.2.0/gems/addressable-2.4.0/lib/addressable/uri.rb:792:in `initialize'
        from /usr/local/lib/ruby/gems/2.2.0/gems/addressable-2.4.0/lib/addressable/uri.rb:135:in `new'
        from /usr/local/lib/ruby/gems/2.2.0/gems/addressable-2.4.0/lib/addressable/uri.rb:135:in `parse'
        from /usr/local/lib/ruby/gems/2.2.0/gems/travis-1.8.0/lib/travis/cli/repo_command.rb:75:in `detect_slug'
        from /usr/local/lib/ruby/gems/2.2.0/gems/travis-1.8.0/lib/travis/cli/repo_command.rb:60:in `find_slug'
        from /usr/local/lib/ruby/gems/2.2.0/gems/travis-1.8.0/lib/travis/cli/repo_command.rb:21:in `setup'
        from /usr/local/lib/ruby/gems/2.2.0/gems/travis-1.8.0/lib/travis/cli/command.rb:197:in `execute'
        from /usr/local/lib/ruby/gems/2.2.0/gems/travis-1.8.0/lib/travis/cli.rb:64:in `run'
        from /usr/local/lib/ruby/gems/2.2.0/gems/travis-1.8.0/bin/travis:18:in `<top (required)>'
        from /usr/local/bin/travis:23:in `load'
        from /usr/local/bin/travis:23:in `<main>'
```

### PR DESCRIPTION
**Pull request by grim-luminal**
Replace addressable, which doesn't support ssh/git style URIs with a
combination of URI and URI::GitSsh to avoid errors when detecting the
slug from remotes.

Fixes #342

### CODE REVIEW

**Comment by grim-luminal** [comment]
Looks like there's still issues to work out with Ruby 1.8.7 here.

**Comment by grim-luminal** [comment]
The build is now passing for ruby 1.8.7 (as well as everything else).

**Comment by simplyianm** [comment]
:+1:

 this error is annoying

File: travis.gemspec [hunk 1]
```diff
@@ -299,7 +299,9 @@ Gem::Specification.new do |s|
   s.add_dependency "launchy",               "~> 2.1"
   s.add_dependency "typhoeus",              "~> 0.6", ">= 0.6.8"
   s.add_dependency "pusher-client",         "~> 0.4"
-  s.add_dependency "addressable",           "~> 2.3"
+  if Gem::Version.new(RUBY_VERSION) < Gem::Version.new("1.9.0")
```
Comment (BanzaiMan) [inline travis.gemspec]: This doesn't seem necessary, as we require Ruby 1.9.3 now.

File: lib/travis/cli/repo_command.rb [hunk 2]
```diff
@@ -83,6 +82,14 @@ def detect_slug
           end
         end
 
+        def parse_remote(url)
+          URI.parse(url).path
+        rescue URI::InvalidURIError
```
Comment (BanzaiMan) [inline lib/travis/cli/repo_command.rb]: I am against using exceptions as a part of the flow control. If `git@…` is an expected "URI", we should handle it as such.

**Comment by BanzaiMan** [comment]
Also, there should be specs.

File: travis.gemspec (hunk 1 above)
Comment (BanzaiMan) [inline travis.gemspec]: Ah, I see. We are still running builds on 1.8.7. We should probably fix that.

Comment (grim-luminal) [inline travis.gemspec]: >  Ah, I see. We are still running builds on 1.8.7. We should probably fix that.

Nice. 1.9.3 or greater makes things much easier. I can add that change in this PR if it makes sense (any other PR is going to fail the build until the issue with addressable is resolved anyways).

**Comment by grim-luminal** [comment]
> Also, there should be specs.

I totally agree. I was looking for a good way to do that, but as things stand this would require writing specs for a private method, which I can do, but it seems somewhat odd. I did look into extracting the Git remote URI handling code out to it's own module under `Travis::Tools` and writing specs for that but I wasn't sure if Tools was the right place to put it.

File: travis.gemspec (hunk 1 above)
Comment (BanzaiMan) [inline travis.gemspec]: That's already taken care of on the `master` branch. You'd have to update this PR.

Comment (grim-luminal) [inline travis.gemspec]: Thanks, this should be fixed now.

File: lib/travis/cli/repo_command.rb (hunk 2 above)
Comment (grim-luminal) [inline lib/travis/cli/repo_command.rb]: This has been updated to explicitly handle `git@github.com` remotes instead of relying on the `InvalidURIError`.

**Comment by grim-luminal** [comment]
@BanzaiMan I added some specs and addressed your other concerns. The specs seem weird to me and I'm totally open to doing this a different way if you any suggestions as to where to move the remote parsing stuff to make it easier to test.

File: lib/travis/cli/repo_command.rb [hunk 3]
```diff
@@ -83,6 +82,16 @@ def detect_slug
           end
         end
 
+        def parse_remote(url)
+          if url.start_with?('git@github.com:')
```
Comment (BanzaiMan) [inline lib/travis/cli/repo_command.rb]: I just noticed that this assumes that the remote repo resides on github.com. This is true for the most part, but we also have Travis Enterprise, which works with GitHub Enterprise. From what I understand, the only thing we know for certain is that the user name is `git`.

Perhaps
```ruby
if url =~ /^git@[^:]+:/
⋮
```
is reasonably close (if not completely accurate).

**Comment by grim-luminal** [comment]
Updated to handle Github Enterprise URIs.

**Comment by BanzaiMan** [comment]
OK. I will release a 1.8.1 tomorrow.

**Pull request merged**